### PR TITLE
Added support for Razer Blade 15 (2020) Advanced

### LINKED ISSUE #1159: Support for Razer Blade 15 Advanced Model (Early 2020, with OLED display)
### Device Info

Name was not found under the device. Since it did not appear there, I grabbed it from the sticker on the box instead.

* Name: Razer Blade 15 Advanced Model [Early 2020 model]
* Product Number: RZ09-03305E53
* Model Number: RZ09-0330

### Device Information
```
Please post the output of
    lsusb -d 1532: -v
```

```
Bus 001 Device 003: ID 1532:0253 Razer USA, Ltd Razer Blade
Device Descriptor:
  bLength                18
  bDescriptorType         1
  bcdUSB               2.00
  bDeviceClass            0
  bDeviceSubClass         0
  bDeviceProtocol         0
  bMaxPacketSize0        64
  idVendor           0x1532 Razer USA, Ltd
  idProduct          0x0253
  bcdDevice            2.00
  iManufacturer           1 Razer
  iProduct                2 Razer Blade
  iSerial                 0
  bNumConfigurations      1
  Configuration Descriptor:
    bLength                 9
    bDescriptorType         2
    wTotalLength       0x0054
    bNumInterfaces          3
    bConfigurationValue     1
    iConfiguration          0
    bmAttributes         0xa0
      (Bus Powered)
      Remote Wakeup
    MaxPower              500mA
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        0
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      1 Boot Interface Subclass
      bInterfaceProtocol      1 Keyboard
      iInterface              0
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      61
         Report Descriptors:
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x81  EP 1 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0008  1x 8 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        1
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0
      bInterfaceProtocol      1 Keyboard
      iInterface              0
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength     159
         Report Descriptors:
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x82  EP 2 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0010  1x 16 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        2
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0
      bInterfaceProtocol      2 Mouse
      iInterface              0
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      94
         Report Descriptors:
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x83  EP 3 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0008  1x 8 bytes
        bInterval               1
Device Status:     0x0000
  (Bus Powered)
```

```
Please post the output of
    ls -lah /dev/input/by-id/
```

```
total 0
drwxr-xr-x 2 root root 200 May 31 10:44 .
drwxr-xr-x 4 root root 880 May 31 11:04 ..
lrwxrwxrwx 1 root root  10 May 31 10:44 usb-Azurewave_Integrated_Camera_200901010001-event-if00 -> ../event11
lrwxrwxrwx 1 root root  10 May 31 10:44 usb-Azurewave_Integrated_Camera_200901010001-event-if02 -> ../event12
lrwxrwxrwx 1 root root  10 May 31 10:44 usb-ELAN_Touchscreen-event-if00 -> ../event30
lrwxrwxrwx 1 root root  10 May 31 10:44 usb-Razer_Razer_Blade-event-if01 -> ../event26
lrwxrwxrwx 1 root root  10 May 31 10:44 usb-Razer_Razer_Blade-event-kbd -> ../event24
lrwxrwxrwx 1 root root  10 May 31 10:44 usb-Razer_Razer_Blade-if01-event-kbd -> ../event25
lrwxrwxrwx 1 root root  10 May 31 10:44 usb-Razer_Razer_Blade-if02-event-mouse -> ../event29
lrwxrwxrwx 1 root root   9 May 31 10:44 usb-Razer_Razer_Blade-if02-mouse -> ../mouse0
```

### Packet Captures

<Please see here as we probably need packet captures>
<https://github.com/openrazer/openrazer/wiki/Reverse-Engineering-USB-Protocol>

Will provide upon request if necessary.

### Synapse Screenshots
<Screenshots of all available options as seen on Razer Synapse>

synapse-dashboard

### LINKED ISSUE #1201: Support for Razer Blade 15 2020 Advanced (300hz, 2080 SUPER MQ)
Note, if this is a duplicate of #1159 , let me know and I will close this and move the device info there.

### Device Info

* Name: Not found under device. According to [here](https://support.razer.com/articles/1927452531) it is `Razer Blade 15" Advanced (2020)`
* Product Number: RZ09-03305E43
* Model Number: RZ09-0330

### Device Information
```
$ lsusb -d 1532: -v
Bus 001 Device 004: ID 1532:0253 Razer USA, Ltd Razer Blade
Couldn't open device, some information will be missing
Device Descriptor:
  bLength                18
  bDescriptorType         1
  bcdUSB               2.00
  bDeviceClass            0 
  bDeviceSubClass         0 
  bDeviceProtocol         0 
  bMaxPacketSize0        64
  idVendor           0x1532 Razer USA, Ltd
  idProduct          0x0253 
  bcdDevice            2.00
  iManufacturer           1 
  iProduct                2 
  iSerial                 0 
  bNumConfigurations      1
  Configuration Descriptor:
    bLength                 9
    bDescriptorType         2
    wTotalLength       0x0054
    bNumInterfaces          3
    bConfigurationValue     1
    iConfiguration          0 
    bmAttributes         0xa0
      (Bus Powered)
      Remote Wakeup
    MaxPower              500mA
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        0
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      1 Boot Interface Subclass
      bInterfaceProtocol      1 Keyboard
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      61
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x81  EP 1 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0008  1x 8 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        1
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0 
      bInterfaceProtocol      1 Keyboard
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength     159
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x82  EP 2 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0010  1x 16 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        2
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0 
      bInterfaceProtocol      2 Mouse
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      94
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x83  EP 3 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0008  1x 8 bytes
        bInterval               1
```

```
$ ls -lah /dev/input/by-id/
total 0
drwxr-xr-x 2 root root 180 Jul 16 22:10 .
drwxr-xr-x 4 root root 680 Jul 16 22:10 ..
lrwxrwxrwx 1 root root   9 Jul 16 20:32 usb-Azurewave_Integrated_Camera_200901010001-event-if00 -> ../event4
lrwxrwxrwx 1 root root  10 Jul 16 20:32 usb-Azurewave_Integrated_Camera_200901010001-event-if02 -> ../event18
lrwxrwxrwx 1 root root  10 Jul 16 20:32 usb-Razer_Razer_Blade-event-if01 -> ../event14
lrwxrwxrwx 1 root root  10 Jul 16 20:32 usb-Razer_Razer_Blade-event-kbd -> ../event10
lrwxrwxrwx 1 root root  10 Jul 16 20:32 usb-Razer_Razer_Blade-if01-event-kbd -> ../event11
lrwxrwxrwx 1 root root  10 Jul 16 20:32 usb-Razer_Razer_Blade-if02-event-mouse -> ../event15
lrwxrwxrwx 1 root root   9 Jul 16 20:32 usb-Razer_Razer_Blade-if02-mouse -> ../mouse3
```

### Packet Captures

Captures me starting up Razer's Synapse software, and letting it run for a little bit: https://drive.google.com/file/d/1w2aSdeSqZjMzE4bZPiQJHnARePNA2vTe/view?usp=sharing

### Synapse Screenshots
<Screenshots of all available options as seen on Razer Synapse>

image

image

image

image

### PR DESCRIPTION
**Pull request by TheButlah**
Fixes #1201
Fixes #1159 

Please test this out on your own machine - this is the first time I have done anything like this so its a bit of a shot in the dark for me. I have confirmed that I can echo commands to the daemon, however polychromatic claims that it sees a razer device but that it isn't recognized

Co-authored-by: @eridgraves

### CODE REVIEW

**Comment by TheButlah** [comment]
I've added a packet capture file to #1201.

File: daemon/openrazer_daemon/hardware/keyboards.py
Comment (z3ntu) [inline daemon/openrazer_daemon/hardware/keyboards.py]: Probably `[6, 16]` instead?

File: driver/razerkbd_driver.h
Comment (z3ntu) [inline driver/razerkbd_driver.h:70]: remove the extra newline

**Review by z3ntu — CHANGES_REQUESTED** [review]
about polychromatic not seeing the device, can you check that the daemon loads your device definition (in the daemon log). Then you can also check with the following command whether device was actually detected by the daemon: 
`python3 -c "from openrazer.client import DeviceManager; a = DeviceManager(); print(a.devices)"`

File: daemon/openrazer_daemon/hardware/keyboards.py
Comment (TheButlah) [reply daemon/openrazer_daemon/hardware/keyboards.py]: I counted the keys and there are 6 rows and 15 columns - see the screenshot of the kbd in #1201.

Comment (TheButlah) [reply daemon/openrazer_daemon/hardware/keyboards.py]: I'm not sure why the others use 16 either though - maybe I'm misunderstanding what this number means? I googled a pic of the 2019 version and that one also has 15 columns but uses 16 here

Comment (eridgraves) [reply daemon/openrazer_daemon/hardware/keyboards.py]: Does `MATRIX_DIMS` refer to the keyboard backlights rather than the keys? 

I looked at my Blackwidow Ultimate but am still unsure of where the extra column is coming from. The layout seems to be 6 keys/lights vertical and up to 21 keys/lights horizontal, whereas the code here has `MATRIX_DIMS = [6, 22]` .

Comment (TheButlah) [reply daemon/openrazer_daemon/hardware/keyboards.py]: Each key except the space bar on the Blade 15 advanced has its own backlight, so it shouldn't matter if it is referring to backlights or keys

File: driver/razerkbd_driver.h
Comment (TheButlah) [reply driver/razerkbd_driver.h:70]: fixed

**Comment by TheButlah** [comment]
Tried following [these](https://github.com/openrazer/openrazer/wiki/Setting-up-the-keyboard-driver) instructions, but I keep getting 
```
echo -n '0003:1532:0253.0004' | sudo tee bind > /dev/null
tee: bind: No such device
```
I can successfully unbind 0004 from `hid-generic`, and can re-bind it there, but I cannot seem to bind it in the razerkbd driver. Any ideas?

**Comment by TheButlah** [comment]
@z3ntu I've just attempted to run that command, but it fails with a ModuleImportError. The python driver doesn't appear to be installed. I think whats going on is that I got the kernel module to work, but haven't successfully installed anything else, which is probably why polychromatic fails to detect the device.

In general, I'm totally lost on how to build and install this project from source. I've filed #1213 - could you or another maintainer help provide me some steps on how to get the project installed from source? 

*EDIT*: Thanks for the help in #1213 !

**Comment by TheButlah** [comment]
OK, after following the advice in #1213, here is my current status:

The device seems bound, although maybe more are in there than necessary? I thought only one device should be there, but IDK:
```
❯ pwd
/sys/bus/hid/drivers/razerkbd
❯ ls
0003:1532:0253.0008  0003:1532:0253.0009  0003:1532:0253.000A  bind  module  new_id  uevent  unbind
```
`000A` is the one that corresponds to the "mouse" interface (as per the directions [here](https://github.com/openrazer/openrazer/wiki/Setting-up-the-keyboard-driver#binding-the-chroma-to-the-keyboard-driver)). It appears as though echoing commands to the files inside that folder, such as `matrix_effect_wave` works.

Here is the output of `python3.8 -c "from openrazer.client import DeviceManager; a = DeviceManager(); print(a.devices)"`:
```
[]
```

Polychromatic continues to insist that the daemon doesn't detect the device:
image

Here is the log file that gets generated when stopping plychromatic, deleting the logs, and starting polychromatic and clicking restart daemon:
```
2020-07-25 17:07:49 | razer                          | INFO     | Initialising Daemon (v2.8.0). Pid: 25905
2020-07-25 17:07:49 | razer.screensaver              | INFO     | Initialising DBus Screensaver Monitor
2020-07-25 17:07:49 | razer                          | INFO     | Found device.0: 0003:1532:0253.000A
2020-07-25 17:07:49 | razer                          | CRITICAL | Could not access /sys/devices/pci0000:00/0000:00:14.0/usb1/1-8/1-8:1.2/0003:1532:0253.000A/device_type, file is not owned by plugdev
2020-07-25 17:07:49 | razer                          | INFO     | Serving DBus
```

I appear to be a member of plugdev, so I don't know why its complaining about that:
image

**Comment by TheButlah** [comment]
`ll /sys/devices/pci0000:00/0000:00:14.0/usb1/1-8/1-8:1.2/0003:1532:0253.000A` revealed that all the files were `root:root`. After a reboot, they are now all `root:plugdev`, and polychromatic no longer gives the unrecognized device page. :clinking_glasses: 

Will update as I play around with polychromatic to see if everything works as expected.

**Comment by TheButlah** [comment]
```
❯ python3.8 -c "from openrazer.client import DeviceManager; a = DeviceManager(); print(a.devices)"
[<RazerKeyboard BY2020A99804910>]
```
image
image
image

**Comment by TheButlah** [comment]
I'm surprised that the effects are only controllable from the drop down menu, and not within the polychromatic GUI. Is this expected? I've never used any of the frontend apps before so I have no point of comparison.

The actual keyboard effects from the drop down menu appear to work, including controlling the lighting on the razer logo.

**Comment by TheButlah** [comment]
Just bumped the number of columns from 15 back to 16. Keeping it at 15 caused the ripple effect to not actually hit the column of keys furthest to the right. Weird, because the keyboard definitely has only 15 columns of keys, but /shrug

File: daemon/openrazer_daemon/hardware/keyboards.py
Comment (TheButlah) [reply daemon/openrazer_daemon/hardware/keyboards.py]: Turned out it needed to be 16, otherwise the ripple effect doesn't reach all the way to the right, even though the keyboard only has 15 keys. IDK why, but I've changed it to 16

**Comment by DmitryLoki** [comment]
wow! super great news! dear team, how soon we can see this in production? many thanks in advance! great job!

**Comment by TheButlah** [comment]
@z3ntu Any changes left that need to be made?

**Comment by DmitryLoki** [comment]
yeah )) we need this pr very much :)

**Comment by TheButlah** [comment]
@DmitryLoki it would help the maintainers to merge this PR if you provided feedback on whether it works or not. Eager excitement is wholesome and appreciated, but the best thing to do to help get this merged sooner is to try it out on your machine and let us know if it works :)

**Comment by DmitryLoki** [comment]
> @DmitryLoki it would help the maintainers to merge this PR if you provided feedback on whether it works or not. Eager excitement is wholesome and appreciated, but the best thing to do to help get this merged sooner is to try it out on your machine and let us know if it works :)

yeap everything works fine
image

**Review by z3ntu — APPROVED** [review]
LGTM, thanks a lot!

File: README.md
Comment (z3ntu) [inline README.md]: Actually quick question: Is "Razer Blade 15 (2020) Advanced" a name used for the device somewhere? Most other devices are in the form "Razer Blade 15 Advanced (2020)"

Comment (TheButlah) [reply README.md]: I think thats a mistake on my part

Comment (z3ntu) [reply README.md]: I'll change it to "Razer Blade 15 Advanced (2020)" then - we can still change it later if needed

**Pull request merged**
